core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala

```diff
@@ -217,4 +224,291 @@ class WorkspaceServiceUnitTests extends AnyFlatSpec with OptionValues with Mocki
 
     exception.errorReport.statusCode shouldBe Some(StatusCodes.Unauthorized)
   }
+
+  def mockWsmForAclTests(ownerEmail: String = "owner@example.com",
+                         writerEmail: String = "writer@example.com",
+                         readerEmail: String = "reader@example.com"
+  ): WorkspaceManagerDAO = {
+    val ownerBinding = new RoleBinding().role(IamRole.OWNER).members(List(ownerEmail).asJava)
+    val writerBinding = new RoleBinding().role(IamRole.WRITER).members(List(writerEmail).asJava)
+    val readerBinding = new RoleBinding().role(IamRole.READER).members(List(readerEmail).asJava)
+    val wsmRoleBindings = new RoleBindingList()
+    wsmRoleBindings.addAll(List(ownerBinding, writerBinding, readerBinding).asJava)
+    val wsmDAO = mock[WorkspaceManagerDAO](RETURNS_SMART_NULLS)
+    when(wsmDAO.getRoles(any(), any())).thenReturn(wsmRoleBindings)
+    wsmDAO
+  }
+
+  def mockSamForAclTests(): SamDAO = {
+    val samDAO = mock[SamDAO](RETURNS_SMART_NULLS)
+    when(samDAO.getUserIdInfo(any(), any())).thenReturn(
+      Future.successful(SamDAO.User(UserIdInfo("fake_user_id", "user@example.com", Option("fake_google_subject_id"))))
+    )
+    when(samDAO.getUserStatus(any()))
+      .thenReturn(Future.successful(Option(SamUserStatusResponse("fake_user_id", "user@example.com", true))))
+    samDAO
+  }
+
+  def mockDatasourceForAclTests(workspaceType: WorkspaceType,
+                                workspaceId: UUID = UUID.randomUUID()
+  ): SlickDataSource = {
+    val datasource = mock[SlickDataSource](RETURNS_SMART_NULLS)
+    val googleProjectId = workspaceType match {
+      case WorkspaceType.McWorkspace    => GoogleProjectId("")
+      case WorkspaceType.RawlsWorkspace => GoogleProjectId("fake-project-id")
+    }
+
+    when(datasource.inTransaction[Workspace](any(), any())).thenReturn(
+      Future.successful(
+        Workspace("fake_ns",
+                  "fake_name",
+                  workspaceId.toString,
+                  "fake_bucket",
+                  None,
+                  DateTime.now(),
+                  DateTime.now(),
+                  "creator@example.com",
+                  Map.empty
+        ).copy(workspaceType = workspaceType, googleProjectId = googleProjectId)
+      )
+    )
+    datasource
+  }
+
+  "getAcl" should "fetch policies from Sam for Rawls workspaces" in {
```

Comment:
These tests mostly focus on checking that the right path is followed wrt Sam vs WSM. There are other unit tests in WorkspaceServiceSpec that focus more on the functionality of getAcl and updateAcl

Reply:
There aren't any modifications to `WorkspaceServiceSpec` in this diff. Are you saying the existing coverage in that class is adequate given the additions to `WorkspaceServiceUnitTests` or is there more to be done?

Reply:
They seem pretty comprehensive to me (there's a bunch in `WorkspaceServiceSpec` and `WorkspaceApiServiceSpec`). They are more focused on validation and making sure `ACLUpdate`s get translated to the appropriate policy additions and removals. Some of that logic has been refactored in this PR, but the behavior should be the same.

Example of some of the tests that check all permutations of workspace ACL updates. These **are** using Sam, but they cover some of the more complicated parts of `updateAcl` that are shared between GCP and Azure workspaces like calculating required policy additions and removals - https://github.com/broadinstitute/rawls/blob/92443713d38a3b90500a94585f3f20e9bf9965b5/core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala#L1416

The new behavior that I'm most concerned with testing is the proper routing to Sam or WSM and translating WSM `RoleBindingList` to the equivalent Scala classes. I added a couple role bindings to the mocked WSM responses that should get ignored by Rawls (`DISCOVERER` and `APPLICATION`).